apps/src/applab/Reducers.js

```diff
@@ -0,0 +1,42 @@
+/** @file Redux reducer functions for App Lab.
+ *  @see http://redux.js.org/docs/basics/Reducers.html */
+// Strict linting: Absorb into global config when possible
+/* jshint
+ unused: true,
+ eqeqeq: true,
+ maxlen: 120
+ */
+'use strict';
+
+var ActionType = require('./Actions').ActionType;
+
+function rootReducer(state, action) {
+  if (typeof state === 'undefined') {
+    state = {
+      assetUrl: function () {}
+    };
+  }
+
+  switch (action.type) {
+    case ActionType.SET_LEVEL_PROPS:
+      var newState = $.extend({}, state);
+      [
+        'assetUrl',
+        'isDesignModeHidden',
+        'isEmbedView',
+        'isReadOnlyWorkspace',
+        'isShareView',
+        'isViewDataButtonHidden'
+      ].forEach(function (propName) {
+        if (typeof action.props[propName] !== 'undefined') {
+          newState[propName] = action.props[propName];
+        }
+      });
+      return newState;
```

Comment:
Looks like this comment disappeared from the PR, but I would say #3. Note:
In a world with ES6, we can just do:
var newState = { ...state, _.pick(<prop list>) };

I wonder if really what we want instead of picking off valid properties is
to look for props that aren't in our list, and then throw if we see them.
At that point, we can just do:
var newState = $.extend({}, state, action.props)

On Fri, Mar 4, 2016 at 4:28 PM, Brad Buchanan <notifications@github.com>
wrote:

> In apps/src/applab/Reducers.js
> <https://github.com/code-dot-org/code-dot-org/pull/7148#discussion_r55108757>
> :
>
> > +  switch (action.type) {
> > +    case ActionType.SET_LEVEL_PROPS:
> > +      var newState = $.extend({}, state);
> > +      [
> > +        'assetUrl',
> > +        'isDesignModeHidden',
> > +        'isEmbedView',
> > +        'isReadOnlyWorkspace',
> > +        'isShareView',
> > +        'isViewDataButtonHidden'
> > +      ].forEach(function (propName) {
> > +        if (typeof action.props[propName] !== 'undefined') {
> > +          newState[propName] = action.props[propName];
> > +        }
> > +      });
> > +      return newState;
>
> If I'm going to use _.pick then I might as well use _.assign too, instead
> of $.extend. Neither pick nor assign are currently part of our custom
> lodash build. Do I:
>
>    1. Add pick and assign to our lodash build?
>    2. Just add pick and use with with extend?
>    3. Finally include all of lodash?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/7148/files#r55108757>.
>